### DISCUSSION TITLE
UX/UI Feedback

### DISCUSSION THREAD

**Discussion by CamStLouis** [edwardkort/WWIDesigner#125] [Ideas]
The original title of this post was "Ready to give up," detailing the various ways I have unsuccessfully tried to calibrate the whistle model and have it generate a tonehole layout based remotely in objective reality. After ruining nearly ten feet of tubing over the course of a year trying to achieve my desired result, I found the problem as I documenting all the things I tried.

Two fields are re-used in a way they are not labeled for, in a context that doesn't follow from the steps preceding it. This fact is buried in one of three articles in the documentation concerning the same subject, and so vaguely described that before trying my solution I literally said out loud, "there is NO WAY this could be what's wrong."

The reason I bring this up is because this project is a **phenomenal synthesis of a truly unbelievable diversity of research, and could help the woodwind making community in so many ways.** Especially as the 3D-printing-based historical instrument community continues to grow and seek scientific ways to analyze and describe their instruments, this project could fuel SO much innovation in traditional & historical music. **I want the entire instrument-making community to be talking about this project.**

But seriously guys. It took me, an experienced user of analytical and scientific software, a year and fifty slides of bullet notes and screenshots recording my process to actually use one function of the software for its intended purpose. There are so many unexpected behaviors and labels that don't match the function that's actually performed with the fields that it's like trying to fly a plane via trial-and-error. I have encountered many people who have said, "oh, yeah, I looked into WWID but could never figure out how to use it" in whistle and bagpiping groups.

So, my question is this: if I were to prepare detailed feedback, mock-up screens in a prototyping tool like Figma on how the UX/UI could be improved & simplified, would you folks be in a position to make changes like the folder hierarchy, separating certain tools into two separate GUIs, and relabeling a bunch of stuff? I'm willing to invest significant time into this, and if you folks like the ideas, I'll write updated documentation to cover the changes, but of course that depends on whether you're interested/able to devote that kind of time to this project. 

I don't think any changes to the underlying tool or math are required, just the behavior of some of the GUIs and many of the variable names/labels. I have a few ideas for value-added features, but I will address those separately, and focus just on what can be done with the existing tool. I'd even chip in money to have my friend reskin the interface, make updated graphics and responsive diagrams if that's of interest.

I think with just a few tweaks and some reorganization, this can be a really approachable, helpful tool for craftspeople of all technological competence levels.

**Comment by bhp1** [comment]
Yes, I would be interested in a better user interface.  Thanks for the offer.  UI design has never been my strong suit.  I"m also finding that Java doesn't have the greatest facilities for building user interfaces.  But we made that choice and for now we have to live with it.

Let's start with the issues in your second paragraph.  If you could be more specific about what's wrong, I could see what we can do to address the issues.  Then we could look at the more detailed feedback you're offering to see what we might be able to accomplish.

Given the Java facilities we're working with, though, implementing "updated graphics and responsive diagrams" sound out of our reach for the time being.

**Reply by CamStLouis** [reply]
> Given the Java facilities we're working with, though, implementing "updated graphics and responsive diagrams" sound out of our reach for the time being.

Makes sense, I was just inspired by my 3D "parametric whistle" model; would be cool to have a simple diagram generated that would display the key dimensions (i.e. window geometry & hole position) in each instrument file window rather than having to click "sketch instrument." Anyway, sorry to get sidetracked with feature requests.

> Let's start with the issues in your second paragraph

The issue in this case was the dual usage of the tuning files both as a reference scale for your target frequencies and a place to enter calibration data. In my opinion these functions should be completely separated. Let's look at the table in the documentation:

![image](https://user-images.githubusercontent.com/76916492/113173754-9310bb00-91fe-11eb-9eef-b53d760f98c6.png)

You cannot use a tuning file capable of generating this table for calibration purposes, because if the target frequency is entered, [WWID will interpret it as the nominal playing frequency you can optionally collect](https://github.com/edwardkort/WWIDesigner/wiki/Working-With-the-Whistle-Study-Model#calibrating-the-whistle-model) instead of Fmin/Fmax. 

> Measure minimum and maximum frequencies for each note, and put the values in the "Min Freq" and "Max Freq" columns of the tuning file. In this case, you can _leave the target frequency in the Frequency column_ or leave the Frequency column blank.
**Ok, so the "Frequency" column is ignored during calibration, right?**

> Measure actual playing frequencies, and put the values in the "Frequency" column. In this case, _do not use the tuning file for optimization; use it only for calibrating the model._ Again, do not try to blow notes in tune; aim for the regular increase of air velocity going up the scale that you might use when playing a tune.
**This now says that data in the "Frequency" column is interpreted as observed nominal playing frequency during calibration.** 

> **Which is it?**

In any case, to get the model to calibrate correctly, I had to delete the data in the "Frequency" column. There is no "add/remove frequency" button to do this, so these values must be deleted individually, again, implying that they need not be deleted _en masse._

I don't think this falls under RTFM - essentially this is like saying "the microwave is only a microwave when you come in the front door to your house. If you come in the side door it's a washing machine," so I don't think it's surprising that people are microwaving their socks.

Let's look at the workflow here:

1. Perform a seven-step process to create a reference tuning file for target frequencies.
2. Repeat this process because clicking "finish" closes without saving rather than the expected behavior of opening the tuning file as a tab in WWID, with the option to save from that screen.
3. Add the min/max fields using the "add/remove" button. This button correctly implies that this data is temporary and subject to change. Note that there is no button to remove target frequency data.
4. Input your min/max data from frequency testing.
5. Display the information in a table like the image above to look at where the target frequency sits in the fmin/fmax range, and get an idea of how far off the model is.
6. Delete the target frequencies, again implied as not necessary to delete by lack of a button, and calibrate the model. 
7. Close the tuning file so you can re-open it with your target frequencies.
8. Perform optimizations.

This technically may be due to the model interpreting target frequencies as observed frequencies _only when no min/max frequency data exists, but if so, this behavior isn't specified,_ and I was specifically informed in an earlier thread that capturing min/max data beyond the high A or so was unnecessary. Either way, this is a completely different function from what it's called, and how it's treated by the rest of the program's functionality. Given the seven-step process to building a tuning file, something one wants to build once and repeatedly call for reference, the notion of having to delete the information that took 7 steps to create is quite a leap.

This is just another example of how much potential there is for accumulation of error. Sure, the program works properly if you know _exactly_ what you're doing conceptually at every step of the process, _and_ are performing the exact steps from the documentation rather than what the GUIs imply, _and_ you under understand the implications of the physics to the point that you folks do, but I don't think that's reasonable to expect of the end user. 

Many of the best instrument makers worldwide have only a fundamental understanding of acoustics, and this tool seems to render itself unnecessary by the requirements it makes of the user. If they understood _exactly_ what was going on to the degree necessary, they wouldn't need WWID except maybe as a helpful shortcut for some calculations.

But the tool's called _Woodwind Instrument Designer_, not _Acoustic Physics Calculator_. and if it's to help bring the benefit of all this incredible research to the woodwind makers of the world, I want to help make it match the workflow and UX/UI behaviors that _they_ would expect.

My feedback would be centered around items like the following:

**1. Improving the GUI and file system -** a lot of the GUIs could better represent what actions will be performed. There are a lot of conventions from software design that this would benefit from following, such as reserving the lower right-hand corner of dialogue boxes for "positive/confirm/save" type actions, and the left hand corner for "negative/go back/don't save" operations. It's also a pain to click through a ton of files every time you go between opening a tuning file and an instrument file - have dedicated buttons that always open the same directories, and make the buttons larger.

**2. Simplifying data input -** outsource complexity to the people who need it! There's no reason for a 7-step process to create a tuning file for target frequencies. If people need different symbols, let them configure that somewhere else. Temperament, ratio, and frequency are all redundant. Only one is needed, and I'd recommend Temperament, since most people want to take a given scale and apply either equal temperament, just intonation, or in my case a custom temperament developed by Hans Bracker. Let equal temperament and just intonation be the defaults and make creation of a custom temperament a different wizard.

_And for god's sake get rid of that "Finish" button_. That fucker has gotten me SO many times.

**3. Adding explanatory information -** instrument makers are primarily concerned with "loud/soft" "big/small" "flat/sharp" and "high airflow/low airflow." Even though I _mostly_ understand what I'm looking at on the graph, making adjustments to the model dimensions doesn't always produce the desired results, and it takes quite a bit of trial-and-error to move things where I want them. Especially since the model absolutely hates my headjoint design, being able to display _observed_ Fmin/Fmax data to get an idea of where a target frequency would _really_ fall in the range would be helpful. I had an apparent windway height of 0.001 and the predictions were still badly flat in the upper octave. 

Clarifying what the optimizers are _actually doing_ in practical terms is also essential. Running the hole spacing optimizer often returns a result that the model _knows_ is incorrect, as defined by the predicted value in the table, so what is it optimizing for? I spent this summer developing an algorithm that can take the measurements of any whistle or recorder and suggest dimensions for any key that preserve the tone and playing characteristics of the original, and I still don't know what the fuck this means:

> When optimizing the tuning of more than one note, the optimizer minimizes the sum of the squares of the differences (multiplied by any optimization weights).

That brings me to the next category:

**4. Optimizing for normal usage -** Referring to the above, I haven't naturally found anywhere to enter optimization weights, and there's no indication whether it would be expected to adjust these in normal usage. Honestly, there's no indication at all what normal usage looks like, so putting the functions you would most expect a user to need at the forefront of the interface would help naturally suggest that sort of thing. For example, the FIRST operation you would want to perform is making a new instrument file, and there is literally no way to do this. You have to copy a default file. This kind of thing is what makes people give up before they've even put in any data. If I were using the software for the first time, I would want to see a big, fat button that says "new instrument."

In general, instrument makers would want to do three things:

1. Create a reference file - e.g. tuning, calibration, custom temperament
2. Plan a new instrument - configure fingering patterns (unclear why this is in the tuning file rather than near the hole data table in the main instrument file), choose a temperament and scale that is automatically associated with the instrument file rather than opened independently, etc.
3. Optimize an existing instrument - enter dimensions, select tuning and provide calibration data, and run optimizers.

Again, my intention here isn't just to complain - I would love to help build a UI for this. WWID is capable of absolutely incredible stuff, but this potential just isn't reached by how it's set up right now. Especially given the work being done to redesign and standardize bagpipe reeds based on the acoustic needs of the instrument rather than "what's worked in the past," WWID could fuel incredible innovation in a sector where there are far more bagpipe makers than reedmakers, and the folks researching the bagpipes can't necessarily just go out and acquire 30 years of reedmaking experience to design a reed for a bagpipe that hasn't been played in 500 years. Imagine if they could just enter the dimensions into WWID and get a good idea of where to start!

**Reply by bhp1** [reply]
Ok, one thing at a time ...

> > Measure minimum and maximum frequencies for each note, and put the values in the "Min Freq" and "Max Freq" columns of the tuning file. In this case, you can _leave the target frequency in the Frequency column_ or leave the Frequency column blank.

> **Ok, so the "Frequency" column is ignored during calibration, right?**

Right, _if_ you supply Fmin and Fmax, then we can use them for calibration and can ignore what's in the Frequency column.

> > Measure actual playing frequencies, and put the values in the "Frequency" column. In this case, _do not use the tuning file for optimization; use it only for calibrating the model._ Again, do not try to blow notes in tune; aim for the regular increase of air velocity going up the scale that you might use when playing a tune.

> **This now says that data in the "Frequency" column is interpreted as observed nominal playing frequency during calibration.**

Yes, if you do _not_ supply Fmin and Fmax, and you are calibrating an instrument, then you must be supplying measured data in the Frequency column.

> **Which is it?**

It can be either, one of the other, user's choice.  As it says before the two points you quoted, "You have two choices of what to measure."

> In any case, to get the model to calibrate correctly, I had to delete the data in the "Frequency" column. There is no "add/remove frequency" button to do this, so these values must be deleted individually, again, implying that they need not be deleted _en masse._

It should not have been necessary to delete anything from the column.  If you supplied Fmin and Fmax for all the notes, and got different calibrations with and without data in the Frequency column, that suggests there's a bug in the software.  

Would one of the following UI improvements help?

- Mark each tuning file with a usage type: Target frequencies (for optimization), Measured frequencies (for calibration), or Target frequencies plus measured Fmin and Fmax (for both calibration and optimization).
- Have the calibrator tell you which columns it is using for calibration: Frequency, or Fmin and Fmax.  (The problem is, it makes the decision note by note, depending on what is supplied for that note, not for the tuning as a whole.)

**Reply by CamStLouis** [reply]
The first bullet I think would make a significant improvement; even though I understand 'under the hood' it's the same tool just used for a different purpose, I think making it look like a completely separate thing would help users not accidentally select the optimization file when tuning, and they could be saved in a separate folder. My workflow would be to have a folder for each instrument file, and within that a folder with numbered calibration files referring to different tubes.

**Reply by bhp1** [reply]
> _And for god's sake get rid of that "Finish" button_. That fucker has gotten me SO many times.

Opened Issue #127.

**Comment by bhp1** [comment]
Before we talk about UI improvements, I don't want to lose track of a comment that suggests a problem in functionality, not just in the user interface.

> being able to display observed Fmin/Fmax data to get an idea of where a target frequency would really fall in the range would be helpful. I had an apparent windway height of 0.001 and the predictions were still badly flat in the upper octave.

Are you saying that the Fmax predictions in the second octave were consistently low?  I don't like that.  Would you be willing to share the tuning file, instrument file, and option settings, that you are using so I could investigate?  

I have myself found it useful to graph observed Fmin/Fmax.  I considered doing that in WIDesigner, either by default, or as a user-selected option.  I decided not to, fearing it would complicate things for other users.  The capability is there, it just isn't available to users.

**Reply by CamStLouis** [reply]
Sure, happy to. I dunno if it's the undercut or the convergent-divergent air path, but my whistles really sit higher in the second octave than normal cylindrical whistles. I honestly have no idea how I achieved this but there is nearly no octave disparity. Do you want the file itself or just the data?

**Reply by CamStLouis** [reply]
And yeah the option to graph the actual min/max while keeping WWID's predictions about where the player is naturally likely to blow a note sharp or flat would be helpful. The graphs for both of my whistles incorrectly placed some of the higher octave notes as being out of the min/max range, but in reality they are perfectly achievable. 

It'd also be nice to select from a couple different "breath curves;" since I find the actual airspeed to rise linearly until the G position, then increase only a little until the 2nd octave D, and then increase in a more quadratic way in the second octave. I can benchmark this by adjusting the imaginary line connecting the green dots, but it would probably improve the accuracy of the predicted values if one could select a more realistic breath curve, or better yet, enter an equation describing exactly what they want (again, relegated to some custom configuration panel so as not to confuse new users.)

This would also be helpful on two-octave bagpipes, as the first octave is essentially a flat line jumping straight up to the second octave pressure, which increases very slightly to keep things from dropping into the low octave. This is why pressure-stable drones are so important on this type of bagpipe! Tuning the cavity of the stock to get the drones to phase-lock helps a lot.

**Reply by bhp1** [reply]
> Do you want the file itself or just the data?

The files, please.  You can attach them here, or e-mail to my C&F e-mail.

**Reply by bhp1** [reply]
> It'd also be nice to select from a couple different "breath curves;

Opened [Discussion #128](https://github.com/edwardkort/WWIDesigner/discussions/128).  I think we first need to discuss what real-world breath curves look like.

Breath curves apply only to the flute and whistle models at this time, not to the NAF or reed models.

**Reply by CamStLouis** [reply]
(I haven't forgotten about this - just haven't had time to get the data together)

**Reply by CamStLouis** [reply]
Thanks for your patience - here are the files. With regard to breath curves, I'll see if I can cobble together a water gauge soon. Really behind this week on irritating householdy stuff and it's eating into my instrument making time. Note that in the files here, I deleted the frequency data in the "High_D_HBtradCaldata" file prior to performing optimizations, leaving just the min/max. 

[WwID_Data.zip](https://github.com/edwardkort/WWIDesigner/files/6354452/WwID_Data.zip)

**Reply by bhp1** [reply]
Thanks, Cam.  I must ask for your patience now; it may take a few weeks before I have time to address this.  Meanwhile, can you tell me your approximate altitude above sea level?  I can use that to estimate the air pressure there, which is one of the [air parameters](https://github.com/edwardkort/WWIDesigner/wiki/WIDesigner-Options#air-parameters) in the WIDesigner options.

**Comment by bhp1** [comment]
Your view of the workflow is different from [how we suggest people start out](https://github.com/edwardkort/WWIDesigner/wiki/Using-WIDesigner#starting-out).  Rather than creating a tuning or instrument from scratch, copy one of the instrument files and one of the tuning files delivered with the release, and modify them for your own purposes.  For whistles, sample tuning files include high D (ET and JI), Bb (JI), A (ET and JI), and low D (JI). 

Perhaps we should have a "starting out" wizard that leads new users more deliberately through the steps of setting up the options, an instrument, and a tuning, using the sample files.

**Reply by CamStLouis** [reply]
I think that would be helpful! I think in most cases, an instrument maker would almost always have a design they were already tinkering with, and if it showed promise, would bring it into WWID for further analysis and optimization. Why would I want to open an existing instrument file? It'd be much faster to just enter all my dimensions into a blank file rather than having to delete a bunch of stuff first. Since we frequently work from museum pieces and other makers' work as a starting place, we'll almost always have complete dimensions *and* a custom scale (probably some optimized variant of just intonation for the keys the instrument can reasonably play in) to enter. I really like Hans Bracker's "trad" tuning, for example:

![image](https://user-images.githubusercontent.com/76916492/113499305-c63a9f00-94c9-11eb-9f42-61d5a66e8665.png)

And for bagpipes that play in two octaves, the tuning can get extremely complex. Donald Lindsay has a fascinating and detailed rationale for the tuning of each note on his extended range smallpipes so that nothing sounds too far out when playing in a different key. (The madman has also added individual tuning set screws to his latest model, so that might be moot now, though!)

Anyway, just to demonstrate the power of this software, once I got that error figured out, here's what came out after bumping some stuff around looking at the graph, first try: https://youtu.be/U3PnC5ta-tk. I might have a few tweaks after we see how the CNC mill does verses my clumsy but attentive hand-drilling. 

Seriously though - I don't have a workshop besides my kitchen table, and I can't work by trial-and-error using power tools all day in a small apartment, so having this kind of scientific approach is literally what makes it possible for me to be an instrument maker. What you folks have accomplished is nothing short of genius.

Looking ahead, I have a friendly competition going with Geert Lejune regarding a set of pastoral pipes he measured from the Brussels museum. He has done some incredible innovation on this type of bagpipe, and along with his original measurements, gave me some hints about how he improved them. He changed the fingering from the rather bizarre period fingering to standard continental half-closed, stretched the bore (aka reducing the angle), made some perturbations, and added an octave key among other things, and they sound amazing. 

I am super excited to apply WWID's reedwind analysis tools, especially bore profile and rejiggering the fingering pattern, to this project, and see what I can learn before the one I ordered from him comes next year. My goal is to make world bagpipes and flutes more reliable, more fun to play, and more accessible to young people without so much maintenance and reliance on only a few reedmaking experts around the globe.

**Comment by CamStLouis** [comment]
No worries. Altitude above sea level is maybe... 200 feet max? I live in Seattle so half the time I’m literally playing these on the beach!

Sent via mobile.

**Reply by CamStLouis** [reply]
Let me know if there's any additional data I can provide

**Comment by bhp1** [comment]
Finally got a chance to take a quick look.  I didn't have any better luck than you at calibrating the model against your measured min/max data.  I wonder a bit at the uppermost bore point, 0.1" above the end of the windway.  Is there a hollow underneath the windway?  If so, wouldn't the bore have to be narrower than it is at the end of the windway?  (This bore point doesn't explain the calibration difficulty, though.)

Failing this, let's look at actuals...  How deep would you say the window is?  In other terms, how thick is the head at the window?

The model suggests the whistle, as you've measured it, would be in tune at D5, reasonably in tune if a bit flat through the first octave, then around half a semitone flat through the second octave.  You could blow that second octave in tune, but it would take twice as much air as the first octave.  How does this prediction reflect reality?  Have you, or can you, use RTTA, Flutini, or TTtuner to measure the actual playing frequencies when you aren't actively trying to play the whistle in tune?

**Reply by CamStLouis** [reply]
There is, there's a 0.1" depression beneath the windway face which helps add depth to the tone, at a diameter of around 0.4". The window depth at the sidewall is 0.133," length 0.21" and width 0.32. The whistle is well in tune across the octaves, and only a couple cents flat in the upper octave according to RTTA. I'm quite proud of how these turned out, actually - the volume up to the high G is pretty much the same across the octaves. High A and B start getting louder, of course, but not by much. I don't find myself straining to push the second octave in tune.

**Reply by bhp1** [reply]
So WIDesigner misses the mark in two respects.  It can predict the good tuning across the first octave, but only with the windway height set far below reality.  And its prediction of the first and second octaves don't line up.  Perhaps in some way your mouthpiece design strays outside the population that I used to build the whistle model.

**Reply by CamStLouis** [reply]
I'll send you one when they're in production (shouldn't be too long now!) - might prove interesting if nothing else, and I owe you for your work on this tool. You a D or an Eb man?

**Reply by bhp1** [reply]
Thank you for the offer, Cam.  I'd welcome a chance to improve how we model your whistles.  Key of D would definitely be my choice.  Good luck with your enterprise.

**Comment by CamStLouis** [comment]
Ok! I'm getting the first batch of finals produced soon, I'll follow up
with you once I have them ready to send out!

On Sun, Jun 13, 2021 at 5:49 PM bhp1 ***@***.***> wrote:

> Thank you for the offer, Cam. I'd welcome a chance to improve how we model
> your whistles. Key of D would definitely be my choice.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/edwardkort/WWIDesigner/discussions/125#discussioncomment-866151>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASK2ODEN3ZCGKF2G3W5RHXLTSVGZVANCNFSM42DFD3RA>
> .
>


--
